reps/2024-07-01-raycluster-crd-status-improvement.md

```diff
@@ -0,0 +1,137 @@
+        * If users use Ray Train or Ray Data, these Ray AI libraries provide some application-level fault tolerance. To elaborate, if some Ray tasks/actors on the worker Pods fail, these libraries will launch new Ray tasks/actors to recover the job. In this case, we may consider the RayCluster as `Failed` when the head Pod crashes.
+        * If users use Ray Serve with RayService and enable GCS fault tolerance for high availability, the RayCluster should still be able to serve requests even when the head Pod is down. Should we still consider the RayCluster as `Failed` when the head Pod crashes?
+
+2. Introduce `Status.Conditions` to surface errors with underlying Ray Pods. 
+    * TODO: We still need to define which types of information we want to surface in the CR status, operator logs, or Kubernetes events.
+
+3. Make suspending a RayCluster an atomic operation.
+    * Like RayJob, RayCluster also supports the [suspend](https://github.com/ray-project/kuberay/pull/1711) operation. However, the suspend operation is [atomic](https://github.com/ray-project/kuberay/pull/1798) in RayJob. In RayCluster, if we set suspend to true and then set it back to false before the RayCluster transitions to the state [rayv1.Suspended](https://github.com/ray-project/kuberay/blob/8453f9b6c749abf3b72b09e0d41032bbfcf17367/ray-operator/apis/ray/v1/raycluster_types.go#L115), there might be some side effects.
+
+4. Add a new condition `Status.Conditions[HeadReady]` to indicate whether the Ray head Pod is ready for requests or not. This is also useful for RayJob and RayService, where KubeRay needs to communicate with the Ray head Pod.
+
+5. Redefine `rayv1.Ready`.
+    * As mentioned in the “Context” section, when all Pods in the RayCluster are running for the first time, the RayCluster CR’s `Status.State` will transition to `ready` ([code](https://github.com/ray-project/kuberay/blob/8453f9b6c749abf3b72b09e0d41032bbfcf17367/ray-operator/apis/ray/v1/raycluster_types.go#L113)). It is highly possible that the RayCluster will remain in the `ready` state even if some incidents occur, such as manually deleting the head Pod.
+    * Here, we redefine `ready` as follows:
+        * When all Pods in the RayCluster are running for the first time, the RayCluster CR’s `Status.State` will transition to `ready`.
+        * After the RayCluster CR transitions to `ready`, KubeRay only checks whether the Ray head Pod is ready to determine if the RayCluster is `ready`.
+        * This new definition is mainly to maintain backward compatibility.
+    * TODO
+        * I am considering making `ready` a condition in `Status.Conditions` and removing `rayv1.Ready` from `Status.State`. Kubernetes Pod also makes a [PodCondition](https://pkg.go.dev/k8s.io/api/core/v1#PodConditionType), but this may be a bit aggressive.
+        * In the future, we can expose a way for users to self-define the definition of `ready` ([#1631](https://github.com/ray-project/kuberay/issues/1631)).
+
+## Implementation
```

Comment:
There are some breaking changes, could you talk about why you think it's ok to have those breaking changes

Reply:
@kevin85421 I don't think we need to make any breaking changes right? We can deprecate the existing `.status.state` field and point useers to the new conditions without deleting `.status.state`